src/pybloomfilter.pyx

```diff
@@ -58,122 +57,113 @@ cdef class BloomFilter:
-
-            test = array_seeds.tobytes()
-            seeds = test
-
-            bits_per_hash = math.ceil(
-                    capacity * abs(math.log(error_rate)) /
-                    (num_hashes * (math.log(2) ** 2)))
+                cpython.PyErr_NoMemory()
 
-            # Minimum bit vector of 128 bits
-            num_bits = max(num_hashes * bits_per_hash,128)
+    def _open(self, filename, mode="rw"):
+        # Should not overwrite
+        mode = mode.replace("+", "")
 
-            # print("k = %d  m = %d  n = %d   p ~= %.8f" % (
-            #     num_hashes, num_bits, capacity,
-            #     (1.0 - math.exp(- float(num_hashes) * float(capacity) / num_bits))
-            #     ** num_hashes))
+        if not os.path.exists(filename):
+            raise OSError(eno.ENOENT, '%s: %s' % (os.strerror(eno.ENOENT),
+                                                        filename))
+        if not os.access(filename, _construct_access(mode)):
+            raise OSError("Insufficient permissions for file %s" % filename)
 
-            # If a filename is provided, we should make a mmap-file
-            # backed bloom filter. Otherwise, it will be malloc
-            if filename:
-                self._bf = cbloomfilter.bloomfilter_Create_Mmap(_capacity,
-                                                       error_rate,
-                                                       filename.encode(),
-                                                       num_bits,
-                                                       self._oflags,
-                                                       perm,
-                                                       <int *>seeds,
-                                                       num_hashes)
-            else:
-                self._in_memory = 1
-                self._bf = cbloomfilter.bloomfilter_Create_Malloc(_capacity,
-                                                       error_rate,
-                                                       num_bits,
-                                                       <int *>seeds,
-                                                       num_hashes)
-            if self._bf is NULL:
-                if filename:
-                    raise OSError(errno, '%s: %s' % (os.strerror(errno),
-                                                     filename))
-                else:
-                    cpython.PyErr_NoMemory()
+        self._oflags = _construct_mode(mode)
+        self._bf = cbloomfilter.bloomfilter_Create_Mmap(0,
+                                                0,
+                                                filename.encode(),
+                                                0,
+                                                self._oflags,
+                                                0,
+                                                NULL, 0)
+        if self._bf is NULL:
+            raise ValueError("Invalid %s file: %s" %
```

Comment:
Note that now there's no clear indicator that's equivalent to  `cpython.PyErr_NoMemory()` when heap allocation fails. Should that be added?

Reply:
I see where you're coming from. The original code only had this type of handling for opening existing files. I'm not so familiar with the C API myself, but it would be good to have similar sort of error handling here. I am not sure how to know whether a file contains invalid data or we had an issue during allocation?

Reply:
I see that later commit has fixed this, so lets go ahead with this.
We'll leave a note for later and work on it if the allocation errors actually happen in practice.